Plugin bricks Claude Code - stdin fstat EINVAL crash

# Bug Report: SessionStart hook BRICKS Claude Code - cannot start sessions

## Bug Description

**Severity: CRITICAL** - This bug bricks Claude Code in most directories. After installing claude-mem, Claude Code fails to start new sessions in many/most directories (works in some, like `/tmp`, but fails in others like home directories or project folders). Users often cannot use Claude Code until the plugin is manually uninstalled by editing config files.

The SessionStart hook (`context-hook.js`) crashes immediately on startup with `EINVAL: Invalid argument` when bun attempts to `fstat()` stdin. This prevents Claude Code from starting in any directory where the plugin is active.

The crash occurs in bun's lazy stdin loading (`lazyLoadStream`) before the hook code can even check `stdin.isTTY`. This appears to be caused by Claude Code not providing a valid stdin file descriptor to hooks.

## Steps to Reproduce

1. Install claude-mem: `/plugin marketplace add thedotmack/claude-mem` then `/plugin install claude-mem`
2. Start a new Claude Code session in any directory
3. Observe the SessionStart hook error preventing session startup

## Expected Behavior

The SessionStart hook should gracefully handle cases where stdin is unavailable/invalid and fall back to running without stdin input (similar to the `--colors` flag behavior).

## Error Output

```
SessionStart:startup hook error: Failed with non-blocking status code: 14 |
`,E+=`2. Run: npm run worker:restart
15 | `,E+="3. Restart Claude Code",r&&(E+=`
16 |
17 | If that doesn't work, try: /troubleshoot`),n&&(E=`Worker Error: ${n}
18 |
19 | ${E}`),E}var Q=C.join(z(),".claude","plugins","marketplaces","thedotmack"),It
...
EINVAL: Invalid argument
 syscall: "fstat"
    errno: -22
       at lazyLoadStream (:1:21)
       at getReader (:1:21)
       at ref (:4:29)
       at node:stream:1743:323
       at /home/max/.claude/plugins/cache/thedotmack/claude-mem/9.0.1/scripts/context-hook.js:19:2871
```

## Root Cause Analysis

Looking at the minified `context-hook.js`, the relevant code is:

```javascript
var Et=process.argv.includes("--colors");
if(A.isTTY||Et)
  w(void 0).then(o=>{console.log(o),process.exit(0)});
else{
  let o="";
  A.on("data",t=>o+=t),
  A.on("end",async()=>{...})
}
```

Where `A` is `stdin`. The issue:

1. Claude Code invokes the hook without providing a valid stdin fd
2. Bun's lazy loading tries to `fstat()` stdin to initialize it
3. `fstat()` fails with EINVAL because the fd is invalid/closed
4. This crash happens **before** the `isTTY` check can even run

## Suggested Fix

Wrap stdin access in a try-catch or check for valid fd before accessing stdin properties:

```javascript
let hasValidStdin = false;
try {
  hasValidStdin = process.stdin && process.stdin.fd !== undefined;
  // Or attempt a simple operation that would fail on invalid fd
} catch {
  hasValidStdin = false;
}

if (!hasValidStdin || process.stdin.isTTY || hasColorsFlag) {
  // Run without stdin
} else {
  // Read from stdin
}
```

## Environment

- **Claude-mem version**: 9.0.1
- **Claude Code version**: 2.1.2
- **Bun version**: 1.0.18
- **OS**: Ubuntu 22.04 (Linux 6.8.0-49-generic x86_64)
- **Platform**: Linux

## Logs

Worker logs from `~/.claude-mem/logs/claude-mem-2026-01-09.log`:

```
[2026-01-09 16:49:12.780] [INFO ] [HTTP  ] ← 503 /api/readiness {requestId=GET-1767973752780, duration=0ms}
[2026-01-09 16:49:13.284] [INFO ] [HTTP  ] → GET /api/readiness {requestId=GET-1767973753284}
[2026-01-09 16:49:13.284] [INFO ] [HTTP  ] ← 503 /api/readiness {requestId=GET-1767973753284, duration=0ms}
... (repeated 503 responses to /api/readiness)
[2026-01-09 16:49:25.344] [ERROR] [SYSTEM] Port in use but worker not responding to health checks
```

The worker was returning 503 (not ready) which caused the hook to timeout waiting for it.

## Additional Context

- **The plugin completely bricks Claude Code** - users cannot start new sessions after installation
- Since Claude Code won't start, users cannot run `/plugin uninstall` - they must manually edit:
  - `~/.claude/plugins/installed_plugins.json` (remove the claude-mem entry)
  - `~/.claude/plugins/known_marketplaces.json` (remove the thedotmack entry)
  - Delete `~/.claude/plugins/marketplaces/thedotmack/`
  - Delete `~/.claude/plugins/cache/thedotmack/`
- Multiple concurrent Claude Code sessions may have caused the worker to be in a bad state
- The issue is specific to bun's handling of stdin file descriptors when invoked by Claude Code's hook system
- This is a regression or environment-specific bug - the plugin presumably worked for others before

--- a/plugin/scripts/bun-runner.js
+++ b/plugin/scripts/bun-runner.js
@@ -70,15 +70,55 @@ if (!bunPath) {
   process.exit(1);
 }
 
+// Fix #646: Buffer stdin in Node.js before passing to Bun.
+// On Linux, Bun's libuv calls fstat() on inherited pipe fds and crashes with
+// EINVAL when the pipe comes from Claude Code's hook system. By reading stdin
+// in Node.js first and writing it to a fresh pipe, Bun receives a normal pipe
+// that it can fstat() without errors.
+function collectStdin() {
+  return new Promise((resolve) => {
+    // If stdin is a TTY (interactive), there's no piped data to collect
+    if (process.stdin.isTTY) {
+      resolve(null);
+      return;
+    }
+
+    const chunks = [];
+    process.stdin.on('data', (chunk) => chunks.push(chunk));
+    process.stdin.on('end', () => {
+      resolve(chunks.length > 0 ? Buffer.concat(chunks) : null);
+    });
+    process.stdin.on('error', () => {
+      // stdin may not be readable (e.g. already closed), treat as no data
+      resolve(null);
+    });
+
+    // Safety: if no data arrives within 5s, proceed without stdin
+    setTimeout(() => {
+      process.stdin.removeAllListeners();
+      process.stdin.pause();
+      resolve(chunks.length > 0 ? Buffer.concat(chunks) : null);
+    }, 5000);
+  });
+}
+
+const stdinData = await collectStdin();
+
 // Spawn Bun with the provided script and args
 // Use spawn (not spawnSync) to properly handle stdio
 // Note: Don't use shell mode on Windows - it breaks paths with spaces in usernames
 // Use windowsHide to prevent a visible console window from spawning on Windows
 const child = spawn(bunPath, args, {
-  stdio: 'inherit',
+  stdio: [stdinData ? 'pipe' : 'ignore', 'inherit', 'inherit'],
   windowsHide: true,
   env: process.env
 });
+
+// Write buffered stdin to child's pipe, then close it so the child sees EOF
+if (stdinData && child.stdin) {
+  child.stdin.write(stdinData);
+  child.stdin.end();
+}
 
 child.on('error', (err) => {
   console.error(`Failed to start Bun: ${err.message}`);

